pkg/daemon/update.go

```diff
@@ -556,6 +587,122 @@ func (dn *Daemon) updateKernelArguments(oldConfig, newConfig *mcfgv1.MachineConf
+// Right now it supports default (traditional) and realtime kernel
+func (dn *Daemon) switchKernel(oldConfig, newConfig *mcfgv1.MachineConfig) error {
+	// We support Kernel update only on RHCOS nodes
+	if dn.OperatingSystem != machineConfigDaemonOSRHCOS {
+		return fmt.Errorf("Updating kernel on non-RHCOS nodes is not supported")
+	}
+
+	dn.logSystem("ImageURl  %v - %v ", oldConfig.Spec.OSImageURL, newConfig.Spec.OSImageURL)
+	defaultKernel := []string{"kernel", "kernel-core", "kernel-modules", "kernel-modules-extra"}
+	dn.logSystem("Initiating switch from kernel %v to %v", ValidKernelType(oldConfig.Spec.KernelType), ValidKernelType(newConfig.Spec.KernelType))
+	if ValidKernelType(oldConfig.Spec.KernelType) == realtimeKernelType && ValidKernelType(newConfig.Spec.KernelType) == defaultKernelType {
+		dn.logSystem("Switching to Default kernel now")
+		args := []string{"override", "reset"}
+		args = append(args, defaultKernel...)
+		rtKernelUninstall := []string{"--uninstall", "kernel-rt-core", "--uninstall", "kernel-rt-modules", "--uninstall", "kernel-rt-modules-extra"}
+		args = append(args, rtKernelUninstall...)
+		dn.logSystem("Running rpm-ostree %+q", args)
+		if err := exec.Command("rpm-ostree", args...).Run(); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	var mnt, containerName string
+	var err error
+	if mnt, containerName, err = dn.MountOSContainer(newConfig.Spec.OSImageURL); err != nil {
+		return err
+	}
+
+	defer func() {
+		// Delete container and remove image once we are done with using rpms available in OSContainer
+		podmanRemove(containerName)
+		exec.Command("podman", "rmi", newConfig.Spec.OSImageURL).Run()
+		dn.logSystem("Deleted container and removed OSContainer image")
+	}()
+
+	rtRegex := regexp.MustCompile("kernel-rt(.*).rpm")
+	if ValidKernelType(oldConfig.Spec.KernelType) == defaultKernelType || ValidKernelType(newConfig.Spec.KernelType) == realtimeKernelType {
+		// Switch to RT kernel
+		dn.logSystem("Switching to RT kernel now")
+		args := []string{"override", "remove"}
+		args = append(args, defaultKernel...)
+		rpms, err := ioutil.ReadDir(mnt)
+		if err != nil {
+			return err
+		}
+
+		for _, rpm := range rpms {
+			if rtRegex.MatchString(rpm.Name()) {
+				args = append(args, "--install", fmt.Sprintf("%s/%s", mnt, rpm.Name()))
+			}
+		}
+		cmd := exec.Command("rpm-ostree", args...)
+		dn.logSystem("Command to be executed by rpm-ostree %v", cmd)
+		if err := cmd.Run(); err != nil {
+			return err
+		}
+	}
+
+	if ValidKernelType(oldConfig.Spec.KernelType) == realtimeKernelType && ValidKernelType(newConfig.Spec.KernelType) == realtimeKernelType {
+		if oldConfig.Spec.OSImageURL != newConfig.Spec.OSImageURL {
+			// Update kernel-rt pacakges from latest OSContainer
+			dn.logSystem("Initiating realtime kernel update on host")
+			args := []string{"uninstall", "kernel-rt-core", "kernel-rt-modules", "kernel-rt-modules-extra"}
```

Comment:
Let's at least add

```
// TODO: Compare NEVRAs and avoid doing this if kernel is unchanged
```

(Or if you want, make that optimization now, but not a blocker from my PoV)

Reply:
Also, one thing related to this is that if the set of RPMs ever changes, this will start failing.  And in fact in recent RHEL kernel packages there's some new subpackages.

Hmm.  Probably the most robust thing here is to parse the `requested-base-removals` bit in `rpm-ostree status --json` and match on the same `kernel-rt*` regexp you have above.

That still doesn't handle the case of "kernel-rt grows a dependency on linux-firmware-rt" e.g. but that's harder.

Reply:
Yeah, that's a good concern and things to keep an eye on.
Not sure if parsing requested-base-removals will help here as well. I don't see a direct mapping between packages required for default kernel verses rt-kernel. For example: default kernel has kernel sub-packages installed while we don't install kernel-rt sub-package for realtime.